File: src/data/courses/dsa/arrays-and-hashing/hashmap-fundamentals/isomorphic-strings/code/cpp/isomorphic_strings_sol1.cpp
class Solution {
public:
    // Two Maps: O(n) — Bidirectional mapping check
    bool isIsomorphic(string s, string t) {
        unordered_map<char, char> s2t, t2s;
        for (int i = 0; i < s.size(); i++) {
            if (s2t.count(s[i]) && s2t[s[i]] != t[i]) return false;
            if (t2s.count(t[i]) && t2s[t[i]] != s[i]) return false;
            s2t[s[i]] = t[i];
            t2s[t[i]] = s[i];
        }
        return true;
    }
};

File: src/data/courses/dsa/arrays-and-hashing/hashmap-fundamentals/isomorphic-strings/code/python/isomorphic_strings_sol1.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        # Two Maps: O(n) — Bidirectional mapping check
        s2t, t2s = {}, {}
        for sc, tc in zip(s, t):
            if sc in s2t and s2t[sc] != tc: return False
            if tc in t2s and t2s[tc] != sc: return False
            s2t[sc] = tc; t2s[tc] = sc
        return True